pytorch_lightning/trainer/trainer.py

```diff
@@ -500,20 +500,22 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
         """
         try:
             return trainer_fn(*args, **kwargs)
-        except KeyboardInterrupt:
+        except KeyboardInterrupt as exception:
             rank_zero_warn("Detected KeyboardInterrupt, attempting graceful shutdown...")
             # user could press Ctrl+c many times... only shutdown once
             if not self.interrupted:
                 self.state.status = TrainerStatus.INTERRUPTED
                 self.on_keyboard_interrupt()
-        except BaseException:
+                self.on_exception(exception)
+        except BaseException as exception:
             self.state.status = TrainerStatus.INTERRUPTED
             if distributed_available() and self.world_size > 1:
                 # try syncing remaing processes, kill otherwise
                 self.training_type_plugin.reconciliate_processes(traceback.format_exc())
             self._on_exception()
             # reset bookkeeping
             self.state.stage = None
+            self.on_exception(exception)
```

Comment:
quick question: should the `on_expection raise exception` by default ? so users can prevent this behaviour if they wish.

Reply:
that's a good idea! the only issue I see is currently for KeyboardInterrupt we don't re raise the exception so we would have to add extra logic for that. Or is it ok to raise the exception for KeyboardInterrupt as well? not sure why it's not the case currently

Reply:
For KeyboardInterrupt, I guess it won't really matter as the user know he is triggering the failure.

Reply:
good point. sg will add it!

Reply:
Great !

Reply:
I am running into an issue where if I raise the exception here: https://github.com/PyTorchLightning/pytorch-lightning/pull/9183/files#diff-0b8e28002318107af9756a2140f96a4e6165b5034a6e7ca8a551607ba39f145dR270

if there are multiple callbacks, and the first one uses default `on_exception` it will raise the error, before it can get to the other callbacks which might have a custom `on_exception`. do you know any way around this?

Reply:
Ok, let's investigate this in another PR if we do :)